metapool/sample_sheet.py

```diff
@@ -766,19 +799,19 @@ def func(x):
     def _validate_sample_sheet_metadata(self, metadata):
         msgs = []
 
-        for req in ['Assay', 'Bioinformatics', 'Contact']:
+        # Note: this method is used by all sample sheets, and not all
```

Comment:
I am sorry to be dense, @charles-cowart , but can you say that again in different words?  I'm not getting it :)

Reply:
No problem! Sorry about that - I just read that comment again after a few hours and it is indeed cryptic! 😄 I'm going to walk myself through it again: We have four required sections for a stock Illumina sample-sheet ([Header], [Reads], [Settings], [Data]) and KLSampleSheets require two additional sections ([Bioinformatics], [Contact]). Ideally the validation process checks for the presence of all six and raises an Error if one is missing or unrecognized.

We don’t usually expect a section to be missing but it has happened a few times. We don’t expect a section with a completely random name but this would cover the case for misspelled section names.

The new and optional [SampleContext] section changes things up in that it shouldn’t be required for older sheet-types, but should be required for KLSampleSheetWithSampleContext and its children. We therefore need to allow the possibility of section names other than the first six in the loop that parses the sample-sheet.

If we encounter something other than the required six sections during validation (or even during parsing), we need to confirm that it’s a valid optional section ([SampleContext]) or raise an Error if it’s anything else ([Charlie’sArbitrarySection]). That was my thinking, I believe. Also, since the sheet is parsed line by line and section by section, I'd be concerned if an arbitrarily large section containing only junk could be read in and appended to a KLSampleSheet() object; I haven't checked that code recently for this though so I might just be misremembering.

Reply:
If I understand you, for the validation step, this is already accounted for here: 

https://github.com/biocore/metagenomics_pooling_notebook/blob/d753713f6520ca8a5688ab990e7868030796b81a/metapool/sample_sheet.py#L812-L818

... and then tested here: 

https://github.com/biocore/metagenomics_pooling_notebook/blob/d753713f6520ca8a5688ab990e7868030796b81a/metapool/tests/test_sample_sheet.py#L524-L530

In *parsing*, I agree that there is no handling for this.  However, I am not going anywhere near adding additional validation into the parsing function in this PR.  I did something as simple as the below, which added a check to make sure that the expected columns were in each "extra" dataframe section being loaded, and _**all h%ll**_ broke loose in the test suite ... there are a really large number of test files out there that do NOT have the `contains_replicates` column in the `Bioinformatics` section but are being used to instantiate v100 sample sheets (which should be invalid).  However, we can't just change them all over to include the extra column required by v100, because some of them are also being used, in other tests, to instantiate v90 sample sheets, which do not allow this column -_-  Thus, I conclude that adding more validation to _parse would be a big job in its own right.

```
                elif section_name in self.KL_ADDTL_DF_SECTIONS:
                    if getattr(self, section_name) is not None:
                        # vals beyond the header are empty values so don't add
                        # them
                        line = line[:len(getattr(self, section_name).columns)]
                        df = getattr(self, section_name)
                        df.loc[len(df)] = line
                    else:
                        # CSV rows are padded to include commas for the longest
                        # line in the file, so we remove them to avoid creating
                        # empty columns
                        col_names = [value for value in line if value != '']

                        ###### ADDED CODE: ######
                        expected_col_names = list(
                            self.KL_ADDTL_DF_SECTIONS[section_name].keys())
                        if set(col_names) != set(expected_col_names):
                            raise ValueError(
                                f"Columns in '{section_name}' section do not "
                                f"match expected columns: {col_names} != "
                                f"{expected_col_names}")
                        ###### END ADDED CODE ######

                        setattr(self, section_name,
                                pd.DataFrame(columns=col_names))
```